library/core/src/str/mod.rs

```diff
@@ -217,6 +216,75 @@ impl str {
         }
     }
 
+    /// Finds the closest `x` not exceeding `index` where `is_char_boundary(x)` is `true`.
+    ///
+    /// This method can help you truncate a string so that it's still valid UTF-8, but doesn't
+    /// exceed a given number of bytes. Note that this is done purely at the character level
+    /// and can still visually split graphemes, even though the underlying characters aren't
+    /// split. For example, the emoji 🧑‍🔬 (scientist) could be split so that the string only
+    /// includes 🧑 (person) instead.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(nearest_char_boundary)]
+    /// let s = "❤️🧡💛💚💙💜";
+    /// assert_eq!(s.len(), 26)
+    /// assert!(!s.is_char_boundary(13));
+    ///
+    /// let closest = s.nearest_char_boundary_max(13);
+    /// assert_eq!(closest, 10);
+    /// assert_eq!(&s[..closest], "❤️🧡");
+    /// ```
+    #[unstable(feature = "nearest_char_boundary", issue = "none")]
+    #[inline]
+    pub fn nearest_char_boundary_max(&self, index: usize) -> usize {
+        if index > self.len() {
+            self.len()
+        } else {
+            let lower_bound = index.saturating_sub(3);
+            let index = self.as_bytes()[lower_bound..=index].iter().rfind(|b| (b as i8) >= -0x40);
+
+            // SAFETY: we know that the character boundary will be within four characters
+            unsafe { index.unwrap_unchecked() }
+        }
+    }
+
+    /// Finds the closest `x` not below `index` where `is_char_boundary(x)` is `true`.
+    ///
+    /// This method is the natural complement to [`neraest_char_boundary_max`]. See that method
+    /// for more details.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index > self.len()`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(nearest_char_boundary)]
+    /// let s = "❤️🧡💛💚💙💜";
+    /// assert_eq!(s.len(), 26)
+    /// assert!(!s.is_char_boundary(13));
+    ///
+    /// let closest = s.nearest_char_boundary_min(13);
+    /// assert_eq!(closest, 14);
+    /// assert_eq!(&s[..closest], "❤️🧡💛");
+    /// ```
+    #[unstable(feature = "nearest_char_boundary", issue = "none")]
+    #[inline]
+    pub fn nearest_char_boundary_min(&self, index: usize) -> usize {
+        if index > self.len() {
+            slice_error_fail(self, index, index)
+        } else {
+            let upper_bound = (index + 4).max(self.len());
```

Comment:
```suggestion
            let upper_bound = (index + 4).min(self.len());
```

Since we want to not exceed the length, right? `.clamp(0, self.len())` might be better, not sure if it optimizes out clamping above zero though.

Reply:
Hmm, you are right about clamping; I'll test that the >= 0 is elided but it should be.

Reply:
It is: https://rust.godbolt.org/z/hcTsahs14